plugins/org.polarsys.capella.scenario.editor.embeddededitor/src/org/polarsys/capella/scenario/editor/embeddededitor/commands/XtextEditorCommands.java

```diff
@@ -57,61 +65,151 @@ public static void xtextToDiagram(Scenario scenario, EmbeddedEditorView embedded
-      // get actors
+      //get actors
       EList<EObject> actors = domainModel.getScenarioType().getParticipants();
-
-      // get messages
-      // TODO ...
-
+      
+      //get messages
+      EList<EObject> messages = domainModel.getMessagesOrReferences();
+      
       Project project;
       BlockArchitecture blockArchitecture;
       project = ProjectExt.getProject(scenario);
       blockArchitecture = BlockArchitectureExt.getBlockArchitecture(Type.SA, project);
-
+      
       doEditingOnActors(scenario, blockArchitecture, actors);
+      
+      doEditingOnMessages(scenario, blockArchitecture, messages);
     }
   }
 
   private static void doEditingOnActors(Scenario scenario, BlockArchitecture blockArchitecture, EList<EObject> actors) {
     // Make sure your element is attached to a resource, otherwise this will return null
     TransactionalEditingDomain domain = TransactionUtil.getEditingDomain(scenario);
     domain.getCommandStack().execute(new RecordingCommand(domain) {
-
+      
       @Override
       protected void doExecute() {
         // Implement your write operations here,
         // for example: set a new name
-        // element.eSet(element.eClass().getEStructuralFeature("name"), "aNewName");
-
+        //element.eSet(element.eClass().getEStructuralFeature("name"), "aNewName");
+        
         InstanceRole instanceRole;
         SystemComponent instance;
         EList<InstanceRole> instanceRoles = scenario.getOwnedInstanceRoles();
-
+        
         for (Iterator<EObject> iterator = actors.iterator(); iterator.hasNext();) {
           EObject actor = iterator.next();
-          instanceRole = InteractionFactory.eINSTANCE.createInstanceRole();// factory.createInstanceRole();
-          instanceRole.setName(((Actor) actor).getName());
-
-          // create Actor = SystemComponent
-          instance = ComponentExt.createSystemActor();
-          instance.setName(((Actor) actor).getName());
-
-          ((SystemAnalysis) blockArchitecture).getOwnedSystemComponentPkg().getOwnedSystemComponents().add(instance);
-
-          Part part = CsFactory.eINSTANCE.createPart(); // csFactory.createPart();
-          part.setAbstractType(instance);
-          part.setName(instanceRole.getName());
-          PartExt.addPart(instance, part, (ComponentArchitecture) blockArchitecture);
-          instanceRole.setRepresentedInstance(part);
-
-          instanceRoles.add(instanceRole);
+          
+          String instanceName = ((Actor)actor).getName();
+          if (instanceRoles.stream().filter(ir -> ir.getName().equals(instanceName)).collect(Collectors.toList()).size() == 0) {
+            instanceRole = InteractionFactory.eINSTANCE.createInstanceRole();
+            instanceRole.setName(((Actor)actor).getName());
+            
+            //search actor by name
+            List<SystemComponent> components = ((SystemAnalysis) blockArchitecture).getOwnedSystemComponentPkg().getOwnedSystemComponents()
```

Comment:
Can also treat the case if is Logical Component, Physical, EPBS layer.
(block architecture)